[4.2] Add namespace support for templates

thednp opened the pull request:
Pull Request for Issue #30272.

### Summary of Changes
As discussed in the #30272 report, Joomla does not support namespaced templates, this commit solves that.

* the `create()` method now checks for namespaced `template` extension types
* the `getNamespaces()` method properly recognize *Administrator* and *Site* templates; in addition, it will look for `templateDetails.xml` file instead of `extensionName.xml` and read the designated namespace (Line 182)

### Testing Instructions
In order to enable namespace for templates, here are the few steps, we'll use the default site template:
* open `site/templates/cassiopeia/templateDetails.xml` and add between `<description>` and `<files>` the following
```markup
<namespace path="src">Joomla\Template\Cassiopeia</namespace>
```
* delete `site/administrator/cache/autoload_psr4.php`
* login to your `site/administrator`
* open the newly created `site/administrator/cache/autoload_psr4.php`, the last element of the return object should be
```
'Joomla\\Template\\Cassiopeia\\Site\\' => [JPATH_SITE . '/templates/cassiopeia/src']
```
* *optional*:  create a custom field in `site/templates/cassiopeia/src/Field` folder and add it to the `templateDetails.xml` along with `addfieldprefix="Joomla\Template\Cassiopeia\Site\Field"` attribute added to the parent `<fieldset>`
* *optional* try and call the newly added custom field from your namespaced template into a module/plugin, test it out


### Actual result BEFORE applying this Pull Request
The `namespacemap.php` class had no support for Site Templates / Administrator Templates. `addfieldpreffix` couldn't be used with template defined custom fields.


### Expected result AFTER applying this Pull Request
Site Templates / Administrator Templates can be developed with namespace features.


### Documentation Changes Required
Yes. (I wish I knew where)

### Questions
* should I commit this change to the `4.1-dev` branch as well?
* should I commit this change to the `staging` branch? If yes, where is the `namespacemap.php` file?

thednp committed:
Add namespace support for templates

Pull Request for Issue #30272.

### Summary of Changes
As discussed in the #30272 report, Joomla does not support namespaced templates, this commit solves that.

* the `create()` method now checks for namespaced `template` extension types
* the `getNamespaces()` method properly recognize *Administrator* and *Site* templates; in addition, it will look for `templateDetails.xml` file instead of `extensionName.xml` and read the designated namespace (Line 182)

**Other change**
* replaced `defined('_JEXEC') or die;` with `defined('JPATH_PLATFORM') or die;`, perhaps this change was also expected? Either way, it was my best guess.

### Testing Instructions
In order to enable namespace for templates, here are the few steps, we'll use the default site template:
* open `site/templates/cassiopeia/templateDetails.xml` and add between `<description>` and `<files>` the following
```markup
<namespace path="src">Joomla\Template\Cassiopeia</namespace>
```
* delete `site/administrator/cache/autoload_psr4.php`
* login to your `site/administrator`
* open the newly created `site/administrator/cache/autoload_psr4.php`, the last element of the return object should be
```
'Joomla\\Template\\Cassiopeia\\Site\\' => [JPATH_SITE . '/templates/cassiopeia/src']
```
* *optional*:  create a custom field in `site/templates/cassiopeia/src/Field` folder and add it to the `templateDetails.xml` along with `addfieldprefix="Joomla\Template\Cassiopeia\Site\Field"` attribute added to the parent `<fieldset>`
* *optional* try and call the newly added custom field from your namespaced template into a module/plugin, test it out


### Actual result BEFORE applying this Pull Request
The `namespacemap.php` class had no support for Site Templates / Administrator Templates. `addfieldpreffix` couldn't be used with template defined custom fields.


### Expected result AFTER applying this Pull Request
Site Templates / Administrator Templates can be developed with namespace features.


### Documentation Changes Required
Yes. (I wish I knew where)

### Questions
* should I commit this change to the `4.1-dev` branch as well?
* should I commit this change to the `staging` branch? If yes, where is the `namespacemap.php` file?

SharkyKZ commented:
This doesn't work well because templates support copy functions.

thednp commented:
@SharkyKZ what do you suggest?

thednp commented:
@SharkyKZ I can confirm, using the template defined custom fields in the options of a plugin doesn't work, so please what do we need to change to make it work? I really think this is important.

SharkyKZ commented:
This works fine, just not when copying templates. If it doesn't work for you at all, the issue must be elsewhere.

thednp commented:
What do you mean "when copying templates" please? You mean multiple template styles? If that's what you mean, I tested that as well, no issue at all.

SharkyKZ commented:
I mean copying the template using `Copy Template` button.

thednp commented:
Well, after a `Copy Template` I'm getting some strange errors:


>JInstaller: :Install: File does not exist `site\tmp\template_copy_5f7581c597a87\Field`
Template Install: Could not copy files from the files source.

>Error installing template
Unable to install new template from temporary folder.

After correcting the folder structure, (I moved `Field` folder inside `src` and updated *templateDetails.xml*) (facepalm) the installation went fine. I think I understand your concern, the new template still keeps the old namespace for it's custom fields and other functionality, if you delete the original template, the new template will throw lots of crits.

I'm now going forward with testing custom fields defined with modules and see how they work across extensions, I suspect it's going to work the same as for the templates given both types have same namespace logic.

SharkyKZ commented:
Modules don't have the same issue because there's no function for copying module extensions.

thednp commented:
Yea, I can confirm, custom fields defined within modules work across extensions. I don't know how to make it work but at this point, I'm happy the template defined fields work inside its back-end panel.

I'm now going forward with testing if `Helper` from `src` folder work properly in template `index.php` and `html` + `html/layouts` overrides. If this test is successful, I think we're good. Would you agree?

SharkyKZ commented:
No. This is not acceptable in its current state. Eventually we could remove copy template functionality completely. But until then I suggest to use a plugin for registering template namespaces.

thednp commented:
In my view adding a plugin to host libraries, helpers, fields and other resources is just like a template framework, which is not the point I'm trying to make. I was initially a template developer, I used to work with all these frameworks, but after some more experience I came to the point where templates can do anything a framework like T3/T4/Helix can do, and without the troubles of maintenance. I want to avoid bloatware and corruptible code as much as possible, clearly Joomla can & should do without.

On the other side, I stand by you, the current state is not acceptable. I really hope we bring back people to Joomla.

thednp committed:
Update namespacemap.php

thednp commented:
@SharkyKZ what is the resolution for this issue?

SharkyKZ commented:
My opinion hasn't changed since the last comment.

thednp commented:
Would be a pity to have Joomla 4 without namespaced templates.

thednp commented:
Please who else has a say in this? 

@brianteeman, @PhilETaylor @eddieajau ?

PhilETaylor commented:
LMAO that you think my opinion counts for anything in joomla's architectural direction. 

The technical reason why namespaced templates is not easy to achieve has been outlined already. 

The only way to implement the copy function is to have the copy function make changes to the namespaces within all the files it copies - which is a recipe for pain as disaster... but not impossible.

thednp commented:
I was looking at commit authors and figured you could have a say in it. Thanks for your input.

zero-24 commented:
Well what holds us back merging what is in this PR and check on the copy feature whether that is a namespaced template and than point to documentation how to copy such templates. I would say this would be a fair compromise what does others think?

thednp commented:
@zero-24 for some reason, the classes defined within namespaced template (mostly fields) cannot autoload in plugins or other extensions, this is, as @SharkyKZ pointed out, related to *copy template* functionality.

zero-24 commented:
> @zero-24 for some reason, the classes defined within namespaced template (mostly fields) cannot autoload in plugins or other extensions, this is, as @SharkyKZ pointed out, related to *copy template* functionality. I can confirm with my tests.

Yes that is the reason i suggested to not allow namespaced templates to be copied than via the GUI. So both ways are possible the only disadvantage for namespaced templates is that there is no simple copy feature anymore from the backend for the reasons pointed out here.

thednp commented:
Sounds good for me. Imma make a great template without any framework, and you will love it :)

SharkyKZ commented:
> What is more important for Joomla: _namespaced templates_ or _copy template_ functionality?

Copy function could be removed completely. Maybe too late for 4.0. But certainly in 5.0.

And templates can already be namespaced, but without using the namespace mapper.

zero-24 commented:
> > What is more important for Joomla: _namespaced templates_ or _copy template_ functionality?
> 
> Copy function could be removed completely. Maybe too late for 4.0. But certainly in 5.0.
> 
> And templates can already be namespaced, but without using the namespace mapper. 

Why should we not allow using the namespace mapper + add that condition to the copy feature?

SharkyKZ commented:
I didn't say we shouldn't. Just clearing the misconception that templates can't be namespaced already.

zero-24 commented:
> I didn't say we shouldn't. Just clearing the misconception that templates can't be namespaced already.

Ok so this (merging here and add a condition to the copy feature) would be a good way to go forward right?

Fedik commented:
I not tried, but I think there should not be problem with copied template, it will continue to use namespace of Source template, that is okau
https://github.com/joomla/joomla-cms/blob/38bbb4ed829bef8e1d7000a5726c23e1f804ceae/libraries/namespacemap.php#L208-L214

or I am wrong?

thednp commented:
@Fedik in my test I can confirm that. Copied templates makes use of `Joomla\Template\originalTemplate` from namespace.

SharkyKZ commented:
It will use whichever is added to namespace map last. Even if it did use the original path, that would still be wrong.

@zero-24 It's fine I guess.

SharkyKZ commented on libraries/namespacemap.php:
```diff
@@ -6,7 +6,7 @@
  */
 
-defined('_JEXEC') or die;
+defined('JPATH_PLATFORM') or die;
```
```suggestion
defined('_JEXEC') or die;
```

thednp committed:
Update libraries/namespacemap.php

Co-authored-by: SharkyKZ <sharkykz@gmail.com>

wilsonge commented:
Potentially in this case the compromise might be delaying the autoloading of templates - what if we only added the template that's being actively loaded on the page by adding it into `HtmlDocument::_loadTemplate`? Then you'll never have duplicates even if templates have the same class names (obviously this is undesirable but.....)

I also like @zero-24 's suggestion of disabling the copy functionality btw if the template is namespaced. With the child template functionality merged and the further stuff @dgrammatiko has planned which should go into 4.1 - i suspect the common usecase for the copy button will be less useful.

dgrammatiko commented:
> i suspect the common usecase for the copy button will be less useful.

Actually that was one of the things which hold me from pushing any changes in the PR for the com_template. Personally I think there **shouldn't** be a copy button mainly for few reasons:
- The core doesn't offer such functionality for any other pieces of code (module, plugin, component)
- It's plain wrong (at least in my pov) to blindly copy someone else's code (even if it's OSS or you've paid for it) through a click of a button.
- With the sub templates/child templates forking a template makes 0 sense

That said maybe we can rethink the namespace (eg in the xml define namespace=foo) so the child templates can extend from the parent namespace (? rough idea, I have no clue what the details could be here, obviously needs some proper research)

> has planned which should go into 4.1

I need to find some time to clean up the PR and write the description (that's gonna be lengthy, it will serve also as the base for the docs)

thednp commented:
I'm glad you guys didn't forget about this, what's up? I'm personally all in with any of your decision guys. What @dgrammatiko explains makes complete sense to me.

thednp commented:
After 3 weeks, what's the verdict of this commit please?

SharkyKZ commented:
I would not go with @wilsonge's suggestion to autoload only active template classes. That would mean having to manually load classes in `com_templates` forms, `com_ajax` (once that is eventually fixed) and anywhere else where we need non-active template. Would be better off with devs using own autoloader in such case.

Blocking copy function in `com_templates` for namespaced templates would be acceptable for me.

wilsonge commented:
OK then let's block copy in namespaced templates then

thednp commented:
Feel free to commit to this branch. Is it possible?

dgrammatiko commented:
> OK then let's block copy in namespaced templates then

@wilsonge can you clarify which templates **could** be namespaced?

What I mean is it will be way easier for me (https://github.com/joomla/joomla-cms/pull/30359 ) to have the namesapaced following only the new, inheritable templates

wilsonge commented:
> @wilsonge can you clarify which templates could be namespaced?

Any templates could be (although I don't think any in core will be). It's more to allow templates with complex PHP logic to autoload their code rather then bundling it in their own index.php

thednp commented:
@wilsonge will this PR get approved in the next beta please?

gostn commented:
pr have not the needed 2 successfully test.

wilsonge commented:
Someone needs to add the code change for disabling clone when this attribute is present. Then it needs two tests. I think it's unlikely this makes it to 4.0 unless that is done very quickly.

thednp commented:
@wilsonge from my experience, Joomla extensions sell the most Joomla, especially templates. If this doesn't go to Joomla 4.0, I don't know who will develop templates with `addfieldpath` and other legacy class standard. I certainly won't, and I love Joomla, I won't let go so easlilly.

So please, get me someone willing to contribute and / or is able to provide guidance on this task, I'm willing to work on it.

thednp commented:
Hi guys @wilsonge @SharkyKZ and anyone else. I'm back to work on Joomla and I've researched the new Beta7 code. 

Probably the easiest way to disable template copy for namespaced templates is to do the following:
* `administrator\components\com_templates\src\View\Template\HtmlView.php` ~line 261 could check for the current template if is namespaced, 
* the code I have in mind

**Method 1**
```php
$btnClass = checkNamespacedExtension($this->template) ? 'btn-primary' : 'btn-primary disabled';
ToolbarHelper::modal('copyModal', 'icon-copy', 'COM_TEMPLATES_BUTTON_COPY_TEMPLATE');
```
This method might require small tooltip for the button itself to let the user know why the action is not possible.

**Method 2**
```php
if ( !checkNamespacedExtension($this->template) )
  ToolbarHelper::modal('copyModal', 'icon-copy', 'COM_TEMPLATES_BUTTON_COPY_TEMPLATE');
}
```
This method is probably better, the template developer can provide documentation on the issue.

Come on guys, lets do this!

The thing is: I don't know how to check if an extension is namespaced without loading the XML file, you guys probably

thednp committed:
Update HtmlView.php

This efectivelly disables the possibility to copy templates if they're namespaced extensions.

thednp committed:
Update HtmlView.php

thednp commented:
Guys, a little help?

richard67 commented:
You need to solve conflicts in that file. Do you need help with that?

thednp commented:
@richard67 yes please

richard67 committed:
Merge branch '4.0-dev' into patch-2

richard67 commented on administrator/components/com_templates/src/View/Template/HtmlView.php:
```diff
@@ -258,7 +258,13 @@ protected function addToolbar()
+				$templateDetails = JPATH_SITE.DIRECTORY_SEPARATOR.'templates'.DIRECTORY_SEPARATOR.($this->template->name).DIRECTORY_SEPARATOR.'templateDetails.xml';
+				$templateXML = simplexml_load_file($templateDetails);
+
+				if ( (string) $templateXML->namespace === null || (string) $templateXML->namespace === '' )
```
```suggestion
				if ((string) $templateXML->namespace === null || (string) $templateXML->namespace === '')
```
Code style

thednp committed:
Update HtmlView.php

richard67 commented:
@thednp I've solved the merge conflict on GitHub. That means if you need to do more changes locally later for some reason, you have to pull changes first from remote so you see them locally.

richard67 commented:
@thednp For namespace stuff I'm not really the expert, so let's wait for feedback from others here.

thednp commented:
@richard67 thank you very much, I wish you stay close to this, it's really important IMO.

In brief: this last change was discussed above and needed to complement the original commit to fully enable namespace for templates. Perhaps you would be kind enough to validate this as soon as bots do their checking, would be so grateful.

Thanks again.

thednp commented:
@wilsonge @dgrammatiko @SharkyKZ @zero-24 and any other reviewer.

As you can see I've implemented all discussed above, however there's room for improvement.

I was thinking lately of a new utility to be added either to the `namespacemap.php` class or the extension parent class, something to return the extension namespace. But since this is not justifiable only for templates, I think a `namespace` object property for all namespaced extensions won't hurt, plus it has the potential to hugely optimize the creation of the namespacemap file.

* we basically store the extension namespace at installation, and just like that you have `$this->template->namespace` similar to `$this->template->name` for instance,
* core extensions might need to add/update this property probably in the db
* when creating the `administrator/autoload_psr4.php` file, it only needs to check this property, no need to load the entire XML file, 

Please let me know what you guys think?

thednp commented:
@SharkyKZ am I to understand that my commits are bad or the above idea about storing the extension namespace into the extension object is bad?

Also, again, what are we waiting for in merging this in?

richard67 commented on administrator/components/com_templates/src/View/Template/HtmlView.php:
```diff
@@ -258,7 +258,13 @@ protected function addToolbar()
+				$templateDetails = JPATH_SITE.DIRECTORY_SEPARATOR.'templates'.DIRECTORY_SEPARATOR.($this->template->name).DIRECTORY_SEPARATOR.'templateDetails.xml';
+				$templateXML = simplexml_load_file($templateDetails);
+
+				if ((string) $templateXML->namespace === null || (string) $templateXML->namespace === '')
```
```suggestion
				if ($templateXML->namespace === null || (string) $templateXML->namespace === '')
```
Maybe it can be simplified even more, but anyway a `(string) $templateXML->namespace === null` seems not to make sense to me.

richard67 commented on administrator/components/com_templates/src/View/Template/HtmlView.php:
```diff
@@ -258,7 +258,13 @@ protected function addToolbar()
 			elseif ($this->type === 'home')
 			{
-				ToolbarHelper::modal('copyModal', 'icon-copy', 'COM_TEMPLATES_BUTTON_COPY_TEMPLATE');
+				$templateDetails = JPATH_SITE.DIRECTORY_SEPARATOR.'templates'.DIRECTORY_SEPARATOR.($this->template->name).DIRECTORY_SEPARATOR.'templateDetails.xml';
```
JPATH_SITE? What if someone copies an admin template?

thednp replied:
You're right, but I'm not sure to how to get the correct link to `templateDetails.xml` file in the context of `com_templates`. Perhaps you have an idea?

richard67 replied:
No, not yet.

thednp replied:
The `JPATH_THEMES` for instance will get you `JPATH_SITE/templates/` when you access it in a SITE template view, or `JPATH_ADMINISTRATOR/templates` when you access it in this context of `com_templates`.

richard67 replied:
That might be the way. You should maybe also look up how we build paths at other places in code, because as far as I remember we don't use `DIRECTORY_SEPARATOR`, we use `/` to concatenate our paths because PHP normalizes paths like this (as well as other languages like e.g. Java do) in this way, so "C:/gaga/somefile.txt" is valid in PHP for "C:\gaga\somefile.txt" on Windows.

thednp replied:
I've improved the code to check for template `client_id` and link the correct `templateDetails.xml`. Please check :)

thednp replied:
In my test, the admin templates show the `COPY TEMPLATE` button since ATUM is not a namespaced extension, while the Cassiopeia won't show the button because I did add a namespace for it for testing purposes.

thednp committed:
Update administrator/components/com_templates/src/View/Template/HtmlView.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

thednp committed:
Update HtmlView.php

As suggested by @richard67, now the check also apply to admin templates.

richard67 commented on administrator/components/com_templates/src/View/Template/HtmlView.php:
```diff
@@ -258,7 +258,12 @@ protected function addToolbar()
 			elseif ($this->type === 'home')
 			{
-				ToolbarHelper::modal('copyModal', 'icon-copy', 'COM_TEMPLATES_BUTTON_COPY_TEMPLATE');
+				$templatesPath = (int) $this->template->client_id === 1 ? JPATH_SITE.'/administrator/templates/' : JPATH_SITE.'/templates/';
```
```suggestion
				$templatesPath = ((int) $this->template->client_id === 1 ? JPATH_ADMINISTRATOR : JPATH_SITE) . '/templates/';
```

richard67 replied:
This should be fixed, too.

thednp replied:
@richard67 won't this be better?

```php
$templatesPath = JPATH_SITE . ((int) $this->template->client_id === 1 ? '/administrator' : '') . '/templates/';
```

richard67 replied:
No. Again: We have the `JPATH_ADMINISTRATOR` variable and so we should use it and not hard-code `administrator` somewhere.

richard67 replied:
And you should have noticed our code style wants a space before and after the `.` for string concatenation.

richard67 commented on administrator/components/com_templates/src/View/Template/HtmlView.php:
```diff
@@ -258,7 +258,12 @@ protected function addToolbar()
 			{
-				ToolbarHelper::modal('copyModal', 'icon-copy', 'COM_TEMPLATES_BUTTON_COPY_TEMPLATE');
+				$templatesPath = (int) $this->template->client_id === 1 ? JPATH_SITE.'/administrator/templates/' : JPATH_SITE.'/templates/';
+				$templateDetails = $templatesPath.($this->template->name).'/templateDetails.xml';
```
```suggestion
				$templateDetails = $templatesPath . $this->template->name . '/templateDetails.xml';
```

thednp replied:
@richard67 what do you think of this, won't it be more appropriate?

```php
$templatesPath = JPATH_SITE . ((int) $this->template->client_id === 1 ? '/administrator' : '') . '/templates/';
```

richard67 replied:
No.

richard67 replied:
We have the `JPATH_ADMINISTRATOR` variable so we should use it and not hard-code `administrator` somewhere.

thednp committed:
Update administrator/components/com_templates/src/View/Template/HtmlView.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

thednp committed:
Update HtmlView.php

richard67 commented on administrator/components/com_templates/src/View/Template/HtmlView.php:
```diff
@@ -258,7 +258,12 @@ protected function addToolbar()
 			elseif ($this->type === 'home')
 			{
-				ToolbarHelper::modal('copyModal', 'icon-copy', 'COM_TEMPLATES_BUTTON_COPY_TEMPLATE');
+				$templatesPath = JPATH_SITE . ((int) $this->template->client_id === 1 ? '/administrator' : '') . '/templates/';
```
Again: We have the JPATH_ADMINISTRATOR variable and so we should use it and not hard-code administrator somewhere.

thednp replied:
```php
$templatesPath =  ((int) $this->template->client_id === 1 ? JPATH_ADMINISTRATOR : JPATH_SITE) . '/templates/';
````

richard67 replied:
Yes, that was what I had suggested, but your code on GitHub still shows the old thing. Maybe you have pushed local changes up before having updated your local branch to the remote?

thednp committed:
Update HtmlView.php

https://github.com/joomla/joomla-cms/pull/30816#discussion_r575666797

richard67 commented on administrator/components/com_templates/src/View/Template/HtmlView.php:
```diff
@@ -258,7 +258,7 @@ protected function addToolbar()
 			elseif ($this->type === 'home')
 			{
-				$templatesPath = JPATH_SITE . ((int) $this->template->client_id === 1 ? '/administrator' : '') . '/templates/';
+				$templatesPath =  ((int) $this->template->client_id === 1 ? JPATH_ADMINISTRATOR : JPATH_SITE) . '/templates/';
```
```suggestion
				$templatesPath = ((int) $this->template->client_id === 1 ? JPATH_ADMINISTRATOR : JPATH_SITE) . '/templates/';
```
No double space, code style.

thednp committed:
Update administrator/components/com_templates/src/View/Template/HtmlView.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

richard67 commented:
I'm just thinking about another way to do it which does not require to read the XML (or from database, if it would go that way) each time:

When installing or updating a 3rd party template, the installer would check the XML if the template is namespaced or not, and set a new column "namespaced" or "nocopy" to either 1 or 0 in the template styles table for that template. For the core templates we would set this in the installation.sql (and the update sql for previous J4 beta versions).

The backend then would only have to read that flag from database together with the other stuff.

richard67 commented:
But anyway I think it needs some kind of decision which way to go.

thednp commented:
Well, I suggested something similar above and got downvoted once, I believe db changes are not desirable. At this point, I believe this might be the right decision.

richard67 commented:
... I mean which way to go with the template copy functionality as such, with respect to child templates.

thednp commented:
Child templates is not going to be implemented afaik. I personally don't mind, the template styles can serve the same purpose to some extent, except template overrides, but those view files can be extended within the template itself without any way to break the layout or anything.

dgrammatiko commented:
> Child templates is not going to be implemented afaik. 

Child templates are already implemented what's missing is the GUI updates and I'm working on them as we speak https://github.com/joomla/joomla-cms/pull/30359

FWIW child templates will not be forkable, there's not need for that in the core

thednp commented:
@dgrammatiko interesting, will that interfere or break the namespaced templates?

dgrammatiko commented:
> interesting, will that interfere or break the namespaced templates?

For the old templates the copy/fork will work as before (assuming that this PR is merged then if there's a namespace field in the XML that will be respected, eg no copy)
For the new mode (parnt-child), there won't be a copy/fork anymore (you can have unlimited childs of a parent template without breaking anything and most important updates won't also break anything)


> the child-theme feature is more like a way to override the layouts in this order:

Yes, the layout overrides but also the static assets overrides as well

thednp commented:
@dgrammatiko would a `namespace` property of the temlate extension help you in any way? 

Me and @richard67 were talking about this.

dgrammatiko commented:
@thednp I was thinking if it would make any sense to consolidate some of the 3 new fields, with child templates we introduced `parent` and `inheritable` and with this, we're introducing `namespace` but I guess they're not really relatable although requiring all the new mode templates to be namespace is something that would be quite logical for me. My 2c

thednp commented:
@dgrammatiko I think child templates will very much like to do `use Joomla\Template\Cassiopeia\Site\SomeClassName`, it will only make more sense to merge this PR right here :D

Fedik commented:
> I'm just thinking about another way to do it which does not require to read the XML

For now it is fine as it is. 
Joomla do not store Namespace anywhere in database for any extension, at least I not found.
However some "helper" to check an extension "Namespace" would be useful in future, I guess. 
But it another story.

richard67 commented:
> Joomla do not store Namespace anywhere in database for any extension, at least I not found.

I didn't think about storing the namespace, I thought about storing a flag telling if the template is namespaced or not, so here would would just check that flag to see if we can copy or not. That flag would be set by the installer.

dgrammatiko commented:
>  I thought about storing a flag telling if the template is namespaced or not

We can assume that the new mode templates are namespaced so we can use the already existing `parent`, `inheritable` flags

richard67 commented:
> We can assume that the new mode templates are namespaced so we can use the already existing `parent`, `inheritable` flags

Can we also assume that the old style templates are **_not_** namespaced?

Fedik commented:
>  I thought about storing a flag telling if the template is namespaced or not

We should not make anything special for template, it just another kind of  extension.
When Joomla will have BlablaHelper::getExtensionNamespace() (or something more smart), that would be enough, and will cover much more than just template.
But as I said, it is another story ;)

dgrammatiko commented:
> Can we also assume that the old mode templates are not namespaced?

Assume was the wrong verb. I meant we can force the new mode to be **only** namespaced and let the legacy part die slowly... The effort is not that much for devs: add the 3 fields `parent`, `inheritable` and `namespace` and also move the assets to the `media` field from the `files`, it's like 5 minutes if you do it manually or a simple script can do that automatically (I'm ignoring here the namespace parts in the PHP files but also that shouldn't be a painful task)

richard67 commented:
> Assume was the wrong verb. I meant we can force the new mode to be **only** namespaced and let the legacy part die slowly...

The point here in this PR is we have to check if we are **_not_** namespaced, and only allow copy then. So as long as the legacy part exists and legacy templates can be namespaced, we can't rely on `parent` and `inheritable` only.

dgrammatiko commented:
> So as long as the legacy part exists and legacy templates can be namespaced, we can't rely on parent and inheritable only.

Right, this is the current approach and it's fine. What I'm trying to say here is that we're already introducing a new mode for templates and it might be wise to make the new mode `namespaced` by default. Also as side effect ignore legacy `namespaced` templates. The reason is way fewer things to break and less code to patch in the next major version (?) where the legacy templates will not be supported anymore

thednp commented:
@SharkyKZ @wilsonge @Fedik @dgrammatiko @infograf768 @richard67 @brianteeman 

Are we RTC yet?

brianteeman commented:
Nothing to do with me I am afraid

alikon commented:
@thednp  before to move a pr on RTC status, we require at least 2 succesfull test

thednp commented:
I know @alikon

Fedik commented:
I have tested this item :white_check_mark: successfully on 1d98ab106f71b6c6287476f012ef7065f3902d50<br /><br />works good for me<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30816">issues.joomla.org/tracker/joomla-cms/30816</a>.</sub>

dgrammatiko commented:
I have tested this item :white_check_mark: successfully on 1d98ab106f71b6c6287476f012ef7065f3902d50<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30816">issues.joomla.org/tracker/joomla-cms/30816</a>.</sub>

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30816">issues.joomla.org/tracker/joomla-cms/30816</a>.</sub>

laoneo commented:
The only input I have is that we have back end and front end templates, so it is a similar situation like with components. So wondering if we should add the client also into the namespace path?

laoneo commented on administrator/components/com_templates/src/View/Template/HtmlView.php:
```diff
@@ -258,7 +258,12 @@ protected function addToolbar()
+				$templatesPath = ((int) $this->template->client_id === 1 ? JPATH_ADMINISTRATOR : JPATH_SITE) . '/templates/';
+				$templateDetails = $templatesPath . $this->template->name . '/templateDetails.xml';
+				$templateXML = simplexml_load_file($templateDetails);
+				if ((string) $templateXML->namespace === '' || $templateXML->namespace === null) {
```
This looks a bit awkward to me, the modal depends on an empty namespace.

thednp replied:
What do you suggest we should change?

laoneo replied:
Don't even know why this should relate to an empty namespace. I have no suggestion, just raising a concern.

thednp replied:
That modal is connected to the copy template feature of the old template system and incompatible with namespace, according to folks who participated in this PR.

thednp replied:
It's true some things don't look very related to the context, like `elseif ($this->type === 'home')`, what's home has to do with copy template?

thednp replied:
@laoneo 

I say we can merge this commit and figure out a better way to handle details later. It doesn't look perfect to everybody but it gets the job done, as with this commit I did what I was asked to do to disable copy template feature for namespaced templates. Me and other devs are waiting for this feature to go through to continue development on other features.

laoneo replied:
Basically with this change here you make "copying a template" deprecated. I would add a message here and push child templates forward then as this is the proper way to go then in the future and not copy a whole template. This needs then also to be documented.

thednp replied:
Can't argue with that. When child templates is done, we should have a clearer picture of what needs to be done with this, since @dgrammatiko is still working on it (child templates) and I think he said it should come by default with namespace support.

Fedik commented:
> So wondering if we should add the client also into the namespace path?

It makes sense. I didn't paid attention but I guess it already should work like that.
I will try to test this case, later.

Fedik commented:
@laoneo Site/Administrator is working:

![Screenshot_2021-02-24_19-02-07](https://user-images.githubusercontent.com/1568198/109036998-e1e4a580-76d2-11eb-848a-d3ef6d55a079.png)

laoneo commented:
Nice. On components I guess the client is after the vendor...

dgrammatiko commented:
> Nice. On components I guess the client is after the vendor...

It would make sense to have `Joomla\\Template\\Site\\Cassiopeia` as it's closer to the media parts: `media/templates/site/cassiopeia` (that's for the new mode, childs) but meh

thednp commented:
@laoneo I only duplicated what the code does for modules, it makes sense to have a proper order like `Joomla\\Template\\Site\\Cassiopeia` for all namespaced templates.

Fedik commented:
> On components I guess the client is after the vendor

it the same,
![image](https://user-images.githubusercontent.com/1568198/109059666-2846fe00-76ed-11eb-962e-c7019baa5202.png)

laoneo commented:
@Fedik thanks for clarification. I guess we are here good to go. But somehow we need a deprecated message that copying a template is not recommended anymore.

Fedik commented:
I think it enough to note somewhere on release page or https://docs.joomla.org/Potential_backward_compatibility_issues_in_Joomla_4#Templates

Something like: "child template" replaces the copy feature, especially for a namespaced template
It requires documentation anyway :wink:

dgrammatiko commented:
> Something like: "child template" replaces the copy feature, especially for a namespaced template.
It requires documentation anyway


We can't really expose child templates without having a GUI in place

Fedik commented:
> We can't really expose child templates without having a GUI in place

Yeah right, I forgot.
Then just note that "copy" will not work for a namespaced template.

rdeutz commented:
I am moving this to 4.1, we are in beta and this is a new feature. Thank you all who have been working on this and bringing it to the state it is now. But if we want to get 4.0 out of the door we need to make a cut.

Side note: this decision has be discussed in the production department and the maintainers team. It's not just mine even if I support the decision.

thednp commented:
I don't understand why it's considered a new feature since J4 was booted with namespace support for all extensions except, interestingly, com_ajax and templates, but anyways. More updates work for the future I guess.

rdeutz commented:
It's a new feature because it adds a new functionality to the system. It also has some side effects e.g. the copy function of the template. The change is too big as we can do this at the state we are now. Hope this explanis it a bit better.

thednp commented:
Yes yes, also the transition from Joomla 1.5 code, finally.

thednp commented:
Guys what's the status of this please?

dgrammatiko commented:
I'm in favor of this but since we're introducing a new mode for templates maybe we should make it a required field for any template that implements child templates and leave the current templates unaffected. What I mean is if someone wants to namespace their template they also need to support child templates (basically one more field in the XML `<inheritable>1</inheritable>` and move all the static assets to `media/templates/(administrator || site)/templateName/`)

Fedik commented:
> maybe we should make it a required field for any template that implements child templates and leave the current templates unaffected

This PR already RTC, 
I think it is better if something new will go separately ;)

dgrammatiko commented:
> I think it is better if something new will go separately, unless it really issue ;)

It's not something new, it's a concern raised before https://github.com/joomla/joomla-cms/pull/30816/files#r582107454 Also probably the reason why this PR is not merged yet.

richard67 commented:
Back to pending due to merge conflicts after rebase.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30816">issues.joomla.org/tracker/joomla-cms/30816</a>.</sub>

richard67 commented:
@thednp We had to rebase the PR to 4.2-dev, which has caused merge conflicts now.

As I can see you have meanwhile deleted the branch for this PR. Does that mean you won't continue this PR and someone else shall take it over? Or are you still available to work on it?

In the latter case you would have to restore the patch-2 branch and update it to latest 4.2-dev. We could help with that if necessary. Or you redo the PR with a new branch for 4.2-dev.

Just let us know if you continue or someone shall take it over.

Thanks in advance.

dgrammatiko commented:
It's a pity that this didn't make it with the new mode templates...

thednp commented:
@richard67 my Joomla PRs, especially the color-picker one, have taken so much of my time I don't even know anymore. @dgrammatiko you referring to child templates? Is this PR crucial for that feature?

dgrammatiko commented:
@thednp child templates don't depend on the namespaces but it would have been great to have them right from the start fro the new templates that support child templates. It will happen at some point because there's a need for namespaces...
Also sorry that the cooler picker is still not merged and the project has to rely on an abandoned/non maintained script for the picker... 🤷‍♂️

thednp commented:
I hope I will have news for you @dgrammatiko soon, I finally found some time and context to learn about custom element slots and I'm planning to release an improved color-picker version to npm (you will also be mentioned), so anyone interested in implementing it in Joomla / Wordpress etc should have a painless transition from what ever color-picker to the possibly best color picker you could ever dream of.

laoneo commented:
The only problem which blocks namespaces for templates is the copy function. Right? Can this be removed with child templates?

dgrammatiko commented:
> Can this be removed with child templates?

You **cannot (by design)** copy neither a parent or a child template, so this was never an issue

laoneo commented:
I mean the whole copy template function
![image](https://user-images.githubusercontent.com/251072/152007669-767d2247-9ecf-4bb6-a70d-6aae0083f06e.png)

dgrammatiko commented:
> I mean the whole copy template function

The button exists only for Legacy templates because nobody wanted to break B/C in a minor. For the templates that either support child templates or are child templates the button is not available (I've updated my previous comment)

dgrammatiko commented:
@laoneo what I'm trying (probably poorly) to say is that the changes in this PR from the `libraries/namespacemap.php` is all it's needed for the templates that support child templates.

laoneo commented:
Partially agree as I would like to see a bootTemplate function which uses the service provider as well.

dgrammatiko commented:
Ok, where do we disagree?

laoneo commented:
> @laoneo what I'm trying (probably poorly) to say is that the changes in this PR from the `libraries/namespacemap.php` is all it's needed for the templates that support child templates.

No disagree, here I would go one step further.

dgrammatiko commented:
> No disagree, here I would go one step further.

Can you explain that step?

dgrammatiko commented:
> To load a template by service provider.

@laoneo could you write some pseudo code for the provider? I struggle to figure out how the boot process would look like (I mean not for a helper, that would be probably the same as it is for modules but for the entry points like index.php, component.php, etc). Maybe I'm missing the file/folder structure that such templates should follow...

HLeithner commented:
Since the original branch has been deleted I recreated the Pull request and updated it to a more proper version at #39011

I'm closing this for now, thanks
